# Diffential Analysis with DESeq2


## Launching the RStudio environment

Once the nf-core/rnaseq pipeline is terminated, the resulting data are stored in the folder results_star_salmon. No we can analyze the results by running DESeq2 on RStudio. First of all, we need to launch it by typing:

```bash
sudo rstudio-server start
```

A pop-up will appear and by clicking on "Open", we will be redirected to the RStudio login page. By inserting the username and the password reported below, we will be able to connect to RStudio.

```bash
Username: gitpod
Password: pass
```

Now we are quite ready to perform the differential expression analysis.


## Differential Expression Analysis

As in all analysis, firstly we need to create a new project:

1) Go to the [File] menu and select [New Project];

2) Select [New Directory], [New Project], name the new directory [/workspace/gitpod/training/DE_analysis/] and click on [Create Project];

3) The new project will be automatically opened in RStudio

We can check whether we are in the correct working directory with [getwd()]. The path ["/workspace/gitpod/training/DE_analysis/"] should appear on your console.

Then, to store our results in an organized way, we will create a folder named [de_results] using the "New Folder" button in the bottom right panel. We will save all our resulting tables and plots in this folder. Next, go to the [File] menu, select [New File], and then [R Script] to create a script editor in which we will save all commands required for the analysis. In the editor type:

```r
## Differential expression analysis with DESeq2
```

and save the file as [de_script.R]. From now on, each command described in the tutorial can be added to your script. The resulting working directory should look like this:

![overview](./img/work_dir_RStudio.png)

The analysis requires several R packages. To utilize them, we need to load the following libraries:

```r
# Loading libraries ----

# tidyverse: a collection of R packages for data manipulation, visualization and modeling
library("tidyverse")
# DESeq2: package for differential gene expression analysis
library("DESeq2")
# pheatmap: a package for creating heatmaps, which will be used to visualize the results
library("pheatmap")
# RColorBrewer: a package for creating color palettes, which will be used to customize the heatmaps
library("RColorBrewer")
```

and the pre-computed DESeq2 object (dds) generated by the nfcore/rnaseq pipeline:

```r
# Import the dds obtained from nfcore/rnaseq ----

load("/workspace/gitpod/training/results_star_salmon/star_salmon/deseq2_qc/deseq2.dds.RData")
```

In DESEq2, the [dds] object is a central data structure that contains the following components: 
- countData: a matrix of raw count data, where each row represents a gene, and each column represents a sample.
- colData: a data frame containing information about the samples, such as the experimental design, treatment, and other relevant metadata.
- design: a formula specifying the experimental design utilized to estimate the dispersion and the log2foldchange.

All these components can be checked with specific command:

```r
# dds inspection ----

head(counts(dds)) # to check the raw counts
colData(dds) # to check the sample info
design(dds) # to check the design formula
```

The inspection of the dds revealed that the colData and the design must be re-organized prior to the analysis. Whit the following command we will rename the column of the colData, we will ensure that the rownames of the metadata are present in the same order as the column names and we reconstruct the colData. Notice that with this operation we also eliminate the sizeFactors already estimated by the [nfcore DESeq2 module](https://github.com/nf-core/rnaseq/blob/master/modules/local/deseq2_qc/main.nf) 

```r
# Creation of metadata starting from the dds colData ----
metadata <- DataFrame(
    sample = colData(dds)$sample,
    condition = colData(dds)$Group1,
    replica = colData(dds)$Group2
)

# Assign names to rows of metadata ----
rownames(metadata) <- colnames(counts(dds))

# Fill the dds colData with the generated metadata ---
colData(dds) <- metadata
```

To avoid errors in DESeq2 is essential to check that sample names match between the colData and the countData, and that the sample are in the correct order:

```r
all(colnames(dds$counts) %in% rownames(metadata)) # Must be TRUE
all(colnames(dds$counts) == rownames(metadata)) # Must be TRUE
```

Now that everything is setted up, we can proceed to generate a new DESeq2 object with the corrected metadata and the right design:

```r
# Creation of a new dds ----

dds_new  <- DESeqDataSet(dds, design = ~ condition)

# dds inspection ----

head(counts(dds_new)) # to check the raw counts
colData(dds_new) # to check the sample info
design(dds_new) # to check the design formula
```

Analyzing the structure of the newly created dds, we can observe the differences:

![overview](./img/dds_comparison.png)

The next step is the determination of the "median ratio" to normalize the data. DESeq2 has a specific function, [estimateSizeFactors], to calculate the size factors. This function is usually called within the [DESeq()] function. We removed the size factors during the reconstruction of the colData, so we need to recalculate them. Assigning the results back to the [dds] object fills the [dds] slot with this information:

```r
# Size factors calculation ----
dds_new <- estimateSizeFactors(dds_new)

sizeFactors(dds_new) # Look at the normalization factors for each sample
```

The normalized counts stored in the [dds] can be expected with the [counts()] function

```r
# Inspect the normalized counts ----

normalized_counts <- counts(dds_new, normalized = TRUE)

# For curiosity, we can also inspect the differences between raw counts and normalized counts ----

head(counts(dds_new, normalized = TRUE))
head(counts(dds_new))
```

The normalized matrix counts can be saved in our results folder:

```r
# Saving normalized counts ----

write.table(normalized_counts, file = "de_results/normalized_counts.txt")
```